src/analysis/output.js

```diff
@@ -0,0 +1,104 @@
+// @flow
+
+/**
+ * This module defines a rich output format for cred scores, so that we can use
+ * it to drive UIs and data analysis.
+ */
+import type {Alias} from "../plugins/identity/alias";
+import type {EdgeWeight} from "../core/weights";
+import type {PluginDeclaration} from "./pluginDeclaration";
+
+// unenforced
+export type Integer = number;
+export type TimestampMs = Integer;
+export type Index = Integer;
+
+/**
+ * Describes an individual node in the contribution graph.
+ * Includes the information in the raw Graph node, along with scoring
+ * information, i.e. its total cred, and the amount of cred that it minted.
+ *
+ * While we are in TimelineCred, the "cred" field will have the total cred,
+ * summed across all time slices. Once we migrate to CredRank, it will just be
+ * the cred directly calculated by the CredRank algorithm.
+ */
+export type OutputNode = {|
+  +address: $ReadOnlyArray<string>,
+  +cred: number,
+  +mint: number,
+  +timestamp: TimestampMs | null,
+  +description: string,
+|};
+
+/**
+ * The first version of the output format will have the nodes, and the plugin
+ * declarations. Including the declarations makes it easy for consumers to do
+ * analyses like: how much cred was minted within each plugin? Within each
+ * contribution type? Etc.
+ */
+export type OutputV1 = {|
+  // Ordered by address
+  +orderdNodes: $ReadOnlyArray<OutputNode>,
```

Comment:
Why is it an array instead of a Map with the address as a key?

Reply:
Several reasons:
- Including the nodes in an array means that every node as an index. Then we can normalize edges and contributors to reference nodes by index, which is much more efficient than referencing by address.
- This is explicitly intended as an output data format. Maps are not JSON-serializable.
- We're storing the addresses as arrays of strings because the raw address format has NUL bytes, and these are ugly when you serialize them in JSON.

Reply:
For serialization problem, can just use a plain object with key/val being address/node

In terms of reference efficiency, key/val lookup on an object is O(1), and in JS the array actually is an Object with keys being 0, 1, 2, etc.

Since the ordering of the nodes in the array doesn't hold any semantic significance, I would argue that its actually less intuitive than key/val object. 

Additionally, these indices will be changing all the time as new nodes are added. Adding just a single node would cause the Index of every other node after it to change.

 In terms of pretty serialization, you could just hash the address string to come up with a "nodeId" instead of a "nodeIndex" as the key. That way if I see a certain nodeId in the serialized JSON I can easily look it up by searching for that NodeID. If it's just an array, there's no easy way to do so, would have to programatically look things up.

If I want to store references to certain NodeIds in other places, its important that the nodeIds be deterministic and based on the actual node address vs changing all the time. e.g. If I want to keep track of certain nodeIds in MetaGame web app database, it would only work if I had some level of guarantee that the nodeId would be consistent. Example use case would be storing each users NodeID in their profile and being able to easily look up their node, or just starting with an address and being able to easily map it to a node without having to know the index in advance. 

Additionally, the references to nodes wouldn't be changing all the time, it would be more "stable", and the git diffs wouldn't be as noisy.

Reply:
> In terms of reference efficiency, key/val lookup on an object is O(1), and in JS the array actually is an Object with keys being 0, 1, 2, etc.

The efficiency I'm worried about is space efficiency. Addresses are quite large, consider "sourcecred/sourcecred/github/PULL/sourcecred/sourcecred/1781" as an example. With this format, the number of addresses we need to write is `N + E`. With your format, we need to write `N + 3E + U`, where `N` is the number of nodes, `E` is the number of edges, and `U` is the number of users. That's why we do the same optimization in the graph serialization format, see: https://github.com/sourcecred/sourcecred/pull/374

Even with just the Output V1 format, the output data is 8.4M uncompressed for the sourcecred project. I believe that a large portion (greater than 50%, probably greater than 90%) of that data is address encoding. In practice the number of edges dominates the number of nodes, so it's not unreasonable to think that the index based format will have half the space requirement of a full address format.

> Since the ordering of the nodes in the array doesn't hold any semantic significance, I would argue that its actually less intuitive than key/val object.

This format is designed to be efficient, not intuitive. As I mentioned somewhere else, we'll make convenient wrapper clients in JS and Python that are intuitive to work with.

> Additionally, these indices will be changing all the time as new nodes are added. Adding just a single node would cause the Index of every other node after it to change.

Adding a new node to the output data is illegal. (To get any meaningful results, you'd need to add the node to the underlying graph, and then re-run cred analysis, which already requires data and context that isn't included in the output format.)

> If I want to store references to certain NodeIds in other places, its important that the nodeIds be deterministic and based on the actual node address vs changing all the time. e.g. If I want to keep track of certain nodeIds in MetaGame web app database, it would only work if I had some level of guarantee that the nodeId would be consistent. Example use case would be storing each users NodeID in their profile and being able to easily look up their node, or just starting with an address and being able to easily map it to a node without having to know the index in advance.

This is what node addresses are for. We don't need a new concept for this.